### ISSUE TITLE
V4.2.4 - remove word "sensitive" from anti-CSRF requirement

### ISSUE THREAD

**Issue by elarlang** [OWASP/ASVS#447]
**requirement**
4.2.4 Verify that the application or framework enforces a strong anti-CSRF mechanism to protect any sensitive functionality.

**recommendation**
I'm not able to give nice and correct requirement text in proper english, but I would like to get rid of the word "sensitive" here. Otherwise we will have discussions (you can call them unnecessary fights) about what is sensitive functionality, how sensitive should functionality be to have CSRF protection etc.

Rhetorical question: what is the functionality for **data operation** where you can skip CSRF protection?

**Comment by jmanico** [comment]
Strong disagree here. Many frameworks do not protect GET’s and similar due to low risk of actual data theft for basic resource requests.

Only sensitive functionality - not basic data retrieval - needs protection.

And if your data retrieval has attribution requirements that is sensitive, then well -protect it.

As is I stand by current language but am happy to debate in more detail.

**Comment by elarlang** [comment]
I agree with GET method and reading data functionality - it's actually does not need CSRF protection and does not belong to this scope, because with CSRF-only you can not read data. 

CSRF scope is about forcing victim's browser to SUBMIT some data to the server, and if you going to do change in data, you must have CSRF defence in place.

If to use HTTP methods ideology, CSRF defence must be in place, when safe HTTP method is not used.
https://en.wikipedia.org/wiki/Hypertext_Transfer_Protocol#Summary_table

**Comment by jmanico** [comment]
Not true. Your public form is POST but CSRF defense is useless there. Some API’s use non standard verbs....

**Comment by elarlang** [comment]
CSRF defence for public form IS useful - I can not force someone's browser to fill the form from another site.

Back to my original "recommendation" section - the problem for me is that the word "sensitive" is not clear, it does not define when CSRF defence should be used and when not.

**Comment by elarlang** [comment]
> Your public form is POST but CSRF defense is useless there.

From this discussion and especially from quoted line I have question - how do you define CSRF?

**Comment by jmanico** [comment]
Let me rephrase. In most situations forcing an anonymous user to POST to a public form via CSRF is low risk. I have seen CSRF attacks against anonymous/unauthenticated features  where shopping carts get pre-filled, default languages are set to KX (Klingon) and similar, but I think CSRF tokens are only required when that form is a certain risk threshold.

Frankly, even GET requests that just read data will require CSRF defense if a user looking at certain data will harm attribution.

Sure conventional wisdom says only protect POST’s and similar, but at the end of day every features does not need CSRF defense. Only ones of a certain RISK level.

Sure, we can wordsmith this. But rules like “all POST requires CSRF” or “no GET require CSRF” or “all unauthenticated forms require CSRF defense” are statements I disagree with. 

But your points well taken. We certainly can wordsmith this better.

**Comment by elarlang** [comment]
I marked "**data operations**" as "bold" in my initial post and I stay with my opinion that all operations which are making changes in data, must be protected against CSRF. It's integrity protection.
So you can leave out form pre-fills.

> ... but I think CSRF tokens are only required when that form is a certain risk threshold.
> Only ones of a certain RISK level.

I share one more opinion because from some perspective it feels a bit slippery way to walk. If we start making statements like "some **data changing** functionality is not critical enough for CSRF token", then people start asking the same question for every ASVS requirement. Maybe we can ignore some headers like CSP, X-XSS-Protection, X-Frame-Options etc for some functionality, because they are not critical enough? I personally really don't want to open that Pandora box.

If you can improve security with small effort and just provide one token which takes away one attack vector, just why to not do that? Especially when every system must have CSRF functionality in place anyway.

> ..  or "no GET require CSRF" or "all unauthenticated forms require CSRF defence" are statements I disagree with.

It's a bit different problem - but if you say that GET may need CSRF protection, then I can argument you misuse GET method and/or there is risk to reveal CSRF token via GET parameter. CSRF token should be considered the same way as sensitive information like session identifier.

**Comment by jmanico** [comment]
Some folks use GET’s for transactions and we need to handle that legacy for CSRF defense which will likely break bookmarks...

And again sometimes just looking up data via CSRF is a huge risk, like CSRF’ing someone to search for illegal content or data above their privilege level via GET.

The edge cases here are significant.

I think “all data changing operations” is a good start but is not complete.

See where I’m going with this, E?

**Comment by elarlang** [comment]
> I think “all data changing operations” is a good start but is not complete.

then "at least" for prefix and it's more valid? :)

> See where I’m going with this, E?

I have shared my opinions but I'm not the one who make decisions here. Thank you for conversation :)

**Comment by jmanico** [comment]
Well I know you’re an AppSec champ. Let me digest all of this and take a swing at clarification. I want to make sure your comments are addressed.

Moar soon.

**Comment by jmanico** [comment]
I surrender. I just dropped the word sensitive via 21228f1 - if this is still an issue please reopen.
